File: facebook/src/main/java/com/facebook/internal/JsonUtil.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.internal;

import android.annotation.SuppressLint;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.*;

/**
 * com.facebook.internal is solely for the use of other packages within the
 * Facebook SDK for Android. Use of any of the classes in this package is
 * unsupported, and they may be modified or removed without warning at any time.
 */
class JsonUtil {
    static void jsonObjectClear(JSONObject jsonObject) {
        @SuppressWarnings("unchecked")
        Iterator<String> keys = (Iterator<String>) jsonObject.keys();
        while (keys.hasNext()) {
            keys.next();
            keys.remove();
        }
    }

    static boolean jsonObjectContainsValue(JSONObject jsonObject, Object value) {
        @SuppressWarnings("unchecked")
        Iterator<String> keys = (Iterator<String>) jsonObject.keys();
        while (keys.hasNext()) {
            Object thisValue = jsonObject.opt(keys.next());
            if (thisValue != null && thisValue.equals(value)) {
                return true;
            }
        }
        return false;
    }

    private final static class JSONObjectEntry implements Map.Entry<String, Object> {
        private final String key;
        private final Object value;

        JSONObjectEntry(String key, Object value) {
            this.key = key;
            this.value = value;
        }

        @SuppressLint("FieldGetter")
        @Override
        public String getKey() {
            return this.key;
        }

        @Override
        public Object getValue() {
            return this.value;
        }

        @Override
        public Object setValue(Object object) {
            throw new UnsupportedOperationException("JSONObjectEntry is immutable");
        }

    }

    static Set<Map.Entry<String, Object>> jsonObjectEntrySet(JSONObject jsonObject) {
        HashSet<Map.Entry<String, Object>> result = new HashSet<Map.Entry<String, Object>>();

        @SuppressWarnings("unchecked")
        Iterator<String> keys = (Iterator<String>) jsonObject.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            Object value = jsonObject.opt(key);
            result.add(new JSONObjectEntry(key, value));
        }

        return result;
    }

    static Set<String> jsonObjectKeySet(JSONObject jsonObject) {
        HashSet<String> result = new HashSet<String>();

        @SuppressWarnings("unchecked")
        Iterator<String> keys = (Iterator<String>) jsonObject.keys();
        while (keys.hasNext()) {
            result.add(keys.next());
        }

        return result;
    }

    static void jsonObjectPutAll(JSONObject jsonObject, Map<String, Object> map) {
        Set<Map.Entry<String, Object>> entrySet = map.entrySet();
        for (Map.Entry<String, Object> entry : entrySet) {
            try {
                jsonObject.putOpt(entry.getKey(), entry.getValue());
            } catch (JSONException e) {
                throw new IllegalArgumentException(e);
            }
        }
    }

    static Collection<Object> jsonObjectValues(JSONObject jsonObject) {
        ArrayList<Object> result = new ArrayList<Object>();

        @SuppressWarnings("unchecked")
        Iterator<String> keys = (Iterator<String>) jsonObject.keys();
        while (keys.hasNext()) {
            result.add(jsonObject.opt(keys.next()));
        }

        return result;
    }
}

File: facebook/src/test/java/com/facebook/internal/JsonUtil.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.internal

import java.lang.IllegalArgumentException
import java.lang.UnsupportedOperationException
import java.util.HashSet
import org.json.JSONException
import org.json.JSONObject

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
internal object JsonUtil {
  fun jsonObjectClear(jsonObject: JSONObject) {
    val keys = jsonObject.keys()
    while (keys.hasNext()) {
      keys.next()
      keys.remove()
    }
  }

  fun jsonObjectContainsValue(jsonObject: JSONObject, value: Any): Boolean {
    val keys = jsonObject.keys()
    while (keys.hasNext()) {
      val thisValue = jsonObject.opt(keys.next())
      if (thisValue != null && thisValue == value) {
        return true
      }
    }
    return false
  }

  fun jsonObjectEntrySet(jsonObject: JSONObject): Set<Map.Entry<String, Any>> {
    val result = HashSet<Map.Entry<String, Any>>()
    val keys = jsonObject.keys()
    while (keys.hasNext()) {
      val key = keys.next()
      val value = jsonObject.get(key)
      result.add(JSONObjectEntry(key, value))
    }
    return result
  }

  fun jsonObjectKeySet(jsonObject: JSONObject): Set<String> {
    val result = HashSet<String>()
    val keys = jsonObject.keys()
    while (keys.hasNext()) {
      result.add(keys.next())
    }
    return result
  }

  fun jsonObjectPutAll(jsonObject: JSONObject, map: Map<String, Any>) {
    val entrySet = map.entries
    for ((key, value) in entrySet) {
      try {
        jsonObject.putOpt(key, value)
      } catch (e: JSONException) {
        throw IllegalArgumentException(e)
      }
    }
  }

  fun jsonObjectValues(jsonObject: JSONObject): Collection<Any> {
    val result = mutableListOf<Any>()
    val keys = jsonObject.keys()
    while (keys.hasNext()) {
      result.add(jsonObject.get(keys.next()))
    }
    return result
  }

  private class JSONObjectEntry constructor(override val key: String, override val value: Any) :
      MutableMap.MutableEntry<String, Any> {
    override fun setValue(newValue: Any): Any {
      throw UnsupportedOperationException("JSONObjectEntry is immutable")
    }
  }
}
